clang/lib/Sema/SemaOverload.cpp

```diff
@@ -11298,8 +11298,9 @@ static void DiagnoseBadConversion(Sema &S, OverloadCandidate *Cand,
   Expr *FromExpr = Conv.Bad.FromExpr;
   QualType FromTy = Conv.Bad.getFromType();
   QualType ToTy = Conv.Bad.getToType();
-  SourceRange ToParamRange =
-      !isObjectArgument ? Fn->getParamDecl(I)->getSourceRange() : SourceRange();
+  SourceRange ToParamRange;
+  if (!isObjectArgument && I < Fn->getNumParams())
+    ToParamRange = Fn->getParamDecl(I)->getSourceRange();
```

Comment:
It seems like the problem here may be that `I` is an index into an argument-as-written list, which doesn't take into account that mapping into the parameter list needs to consider packs.

I think this could do the wrong thing if there are other parameters after the pack.

How does your patch handle the following example?
```C++
template <typename ...a> int b(a..., int);
 int d() { return b<int, int>(0, 0, d); }
```

Reply:
I mean, I would expect the patch would make it so it doesn't crash, but does the diagnostic point to the correct parameter?

Reply:
that's a fair point, I was mostly operating with the assumption of "parameter pack is always the last parameter", but that isn't necessarily true.

as you pointed out this patch prevents the OOB access, and won't necessarily do the "right" thing in terms of  pointing at the failed conversion. but this patch is still an improvement to today's case of just crashing even before printing the candidate.

I think adjusting the parameter range to be accurate in presence of parameter packs is more of a feature request at this point. since (AFAICT) it requires propagating extra information in the bad conversion about the parameter location. so I'd rather move forward with this patch, while filing a feature request for the "ideal" state. WDYT?

Reply:
Sure, fixing the crash is important, but in that case I would leave a FIXME and keep the bug open, with updated information. It's your call.

Though I don't think solving this is complicated: Just a linear scan on the parameter list:
* A parameter which is a pack consumes `cast<PackExpansionType>(Fn->getParamDecl(I)->getType())->getNumExpansions()` arguments, which can possibly be zero arguments. It should not be possible to find an unexpanded pack in this diagnostic.
* A parameter which is not a pack consumes one argument.

Reply:
> Though I don't think solving this is complicated

sorry I guess I didn't convey full context in my previous message when I just said "it requires propagating extra information in the bad conversion about the parameter location".  `Fn` in this context is the template decl itself, hence the packs in the parameters are dependent. hence we can't simply ask how many expansions are there.

Reply:
Okay. Should'nt be a problem to pipe in the specialization, but as I said, just leaving the FIXME is fine.